src/lib/Libutil/misc_utils.c

```diff
@@ -718,60 +718,65 @@ replace(char *str, char	*sub, char *repl, char	*retstr)
 
 /**
  * @brief
- *	Add esc as escape character if delim is appearing on str
+ *	Escape every occurrence of 'delim' in 'str' with 'esc'
  *
- * @param[in]	str		- input string
- * @param[in]	delim  	- delimiter to be searched on str
- * @param[in]   esc 	- escape character to be added if delim found on str
- * @param[out]  retstr 	- delim replaced with and escape character+delim if found.
+ * @param[in]	str     - input string
+ * @param[in]	delim   - delimiter to be searched in str
+ * @param[in]	esc     - escape character to be added if delim found in str
  *
- * @return void
+ * @return	string
+ * @retval	NULL	- memory allocation failed or str is NULL
+ * @retval	retstr	- output string, with every occurrence of delim escaped with 'esc'
  */
 
-void
-escape_delimiter(char *str, char *delim, char esc, char *retstr)
+char *
+escape_delimiter(char *str, char *delim, char esc)
 {
 	int	i=0, j=0;
-	int	delim_len;
-	char	rstr[MAXBUFLEN+1];
+	int	delim_len=0;
+	int retstrlen=0;
+	char *retstr = NULL;
+	char *temp = NULL;
 
 	if (str == NULL)
-		return;
+		return NULL;
 
-	if (*str == '\0') {
-		retstr[0] = '\0';
-		return;
-	}
-
-	if ((delim == NULL || *delim == '\0') || esc == '\0') {
-		strncpy(retstr, str, strlen(str));
-		return;
+	if (*str == '\0' || (delim == NULL || *delim == '\0') || esc == '\0') {
+		return strdup((char *)str);
 	}
 	delim_len = strlen(delim);
+	retstr = (char *)malloc(MAXPATHLEN);
+	if(retstr == NULL)
+		return NULL;
+	retstrlen = MAXPATHLEN;
 
 	while (*str != '\0') {
-		if( *str == esc && i < MAXBUFLEN) {
-				rstr[i++] = esc;
-				rstr[i++] = esc;
+		if(*str == esc) {
+			retstr[i++] = esc;
+			retstr[i++] = esc;
 				str++;
-		} else if(strncmp(str, delim, delim_len) == 0 && i <= MAXBUFLEN) {
-				rstr[i++]=esc;
-			for (j=0; (j < delim_len && i <= MAXBUFLEN); j++, i++) {
-				rstr[i] = delim[j];
+		} else if(strncmp(str, delim, delim_len) == 0) {
```

Comment:
I agree with you that it might have serious performance overhead if the user chooses to have string over char (n*m). I think if strlen(delim) is 1, strncmp would be similar to a character comparison. Should we restrict the user from using delimiter as a string or should both be supported programmatically so that the user can choose to have a string when needed. What do you say?